Documentation on Audited attribute on Entity/Property for EntityHistory (#5126)
aspnetboilerplate/aspnetboilerplate: doc/WebSite/Entity-History.md

```

All properties of MyEntity are tracked except MyProperty2 since it's
explicitly disabled. The **Audited** attribute can be used to
save change logs for a desired property.

> When using **Audited** attribute on **entity** class, **EntityChange will always be created** irregardless of whether any **PropertyChange** is being created. 

Reviewer: `irregardless` -> `regardless`

> Similarly, when using **Audited** attribute on a **property**, **PropertyChange will always be created** irregardless of whether there is any difference between the **new and old values** of the **property**.

Reviewer: This behaviour does not look right.
Unlike `EntityChange`, since the entity did change, in this case it is determined that the property did not change.

Author: This behavior was introduced in https://github.com/aspnetboilerplate/aspnetboilerplate/pull/4870
> mark property as to_be_removed if
> - property'a new value == old value
> - property does not have audited attribute
> - is not a shadow property

Author: i would think the bahavior of `Audited` attribute should be consistent when it is used on either `Entity` or `Property`, meaning changeset will always be created regardless if there is any changes.

If any user wants to create changeset only if `Entity` or `Property` has some changes, they should use `EntityHistory#Selectors`

Reviewer: Currently, it is neither consistent nor correct.

- `EntityChange` is only created if `Entity` has changes<sup>1</sup>. (Correct)
- `PropertyChange` is created regardless of whether `Property` has changes. (Incorrect)

<sup>1</sup> regardless of whether those changes are not recorded in `PropertyChanges`, e.g. due to `DisableAuditing` as in #5093.

Author: > EntityChange is only created if Entity has  

This is not how it behaves at the moment. EntityChannge is always created for entity with Audited attribute

Reviewer: Can you point me to the relevant code for that?

Because `CreateEntityChange` returns `null` in [EntityHistoryHelper.cs#L163-L182 as of v5.1.0](https://github.com/aspnetboilerplate/aspnetboilerplate/blob/v5.1.0/src/Abp.ZeroCore.EntityFrameworkCore/EntityHistory/EntityHistoryHelper.cs#L163-L182):
```c#
[CanBeNull]
private EntityChange CreateEntityChange(EntityEntry entityEntry)
{
    ...
    EntityChangeType changeType;
    switch (entityEntry.State)
    {
        case EntityState.Added:
            changeType = EntityChangeType.Created;
            break;
        case EntityState.Deleted:
            changeType = EntityChangeType.Deleted;
            break;
        case EntityState.Modified:
            changeType = entityEntry.IsDeleted() ? EntityChangeType.Deleted : EntityChangeType.Updated;
            break;
        case EntityState.Detached:
        case EntityState.Unchanged:
            return null;
```

Author: Ahh I think I didn’t elaborate my thoughts well enough.

The code your shown above is valid if Entity does not have any changes at all (no property has any changes). 

However , the scenario that I thought of was when there is a property with changes in the Entity but the property is marked with DisableAuditing attribute.

Reviewer: That's exactly what I said:

> - `EntityChange` is only created if `Entity` has changes<sup>1</sup>. (Correct)
>
> <sup>1</sup> regardless of whether those changes are recorded in `PropertyChanges`, e.g. due to `DisableAuditing` as in #5093.

> Similarly, when using **Audited** attribute on a **property**, **PropertyChange will be created**, regardless of whether there is any differences between the **new and old values** of the **property**.

Reviewer: Still not right.

> As mentioned in https://github.com/aspnetboilerplate/aspnetboilerplate/pull/5126#discussion_r357987933, this behaviour is neither consistent nor correct.
> ...
> Instead, we should fix the regression from #4870.

> - `PropertyChange` should only be created if **property** is changed.

```suggestion
When using **Audited** attribute on a **property**, **PropertyChange will be created** for property if there is any difference between the **new and old values** of the **property**.
```

Reviewer 2: @acjh is this the only change needed ? Or does it still need extra work ?

Reviewer: More work (fix the regression from #4870) is needed to resolve #5292.
Making this change without fixing the regression would make the documentation inconsistent with the behavior.

> Similarly, when using **Audited** attribute on a **property**, **PropertyChange will always be created** for entity in *Addeded/Modified/Deleted* state, regardless of whether there is any differences between the **new and old values** of the **property**.

Reviewer: As mentioned in https://github.com/aspnetboilerplate/aspnetboilerplate/pull/5126#discussion_r357987933, this behaviour is neither consistent nor correct.

This (second) phrasing appears even more forced and does not make any sense:
- `Audited` attribute on **entity** class → `EntityChange` will be created based on **entity** state. (Correct)
- `Audited` attribute on **property** → `PropertyChange` will be created based on **entity** state — what? (Incorrect)

Instead, we should fix the regression from #4870.

Author: > Audited attribute on property → PropertyChange will be created based on entity state — what? (Incorrect)

@acjh can you explain in details?  Since `EntityChange` is only created with entity has changes, and for `PropertyChange` to be created, `EntityChange` need to be created before hand right?

Reviewer: A property change results in an entity change, not the other way around.

A `PropertyChange` itself does not need an `EntityChange` to be created, but is logically accessed only via `EntityChange.PropertyChanges`.

- `EntityChange` should only be created if **entity** is changed.
- `PropertyChange` should only be created if **property** is changed.

`PropertyChange` should **not** be created just because entity has (other) changes.

Reviewer: > **PropertyChange will be created**, regardless of whether there is any differences between the **new and old values** of the **property**.

Still not right. *`PropertyChange` should only be created if property is changed.*
